src/en/asurascans/src/eu/kanade/tachiyomi/extension/en/asurascans/AsuraScans.kt

```diff
@@ -25,47 +54,271 @@ class AsuraScans : MangaThemesiaAlt(
+        val slug = chapter.url.substringAfter("/series/").substringBefore("/")
+        val savedSlug = preferences.slugMap[slug] ?: "$slug-"
+        return baseUrl + chapter.url.replace(slug, savedSlug)
+    }
+
+    override fun pageListRequest(chapter: SChapter): Request {
+        if (!preferences.dynamicUrl()) return super.pageListRequest(chapter)
+        val match = OLD_FORMAT_CHAPTER_REGEX.matches(chapter.url)
+        if (match) throw Exception("Please refresh the chapter list before reading.")
+        val slug = chapter.url.substringAfter("/series/").substringBefore("/")
+        val savedSlug = preferences.slugMap[slug] ?: "$slug-"
+        return GET(baseUrl + chapter.url.replace(slug, savedSlug), headers)
     }
 
-    // Skip scriptPages
     override fun pageListParse(document: Document): List<Page> {
-        return document.select(pageSelector)
-            .filterNot { it.attr("src").isNullOrEmpty() }
-            .mapIndexed { i, img -> Page(i, document.location(), img.attr("abs:src")) }
+        return document.select("div > img[alt=chapter]").mapIndexed { i, element ->
+            Page(i, imageUrl = element.attr("abs:src"))
+        }
+    }
+
+    override fun imageUrlParse(document: Document) = throw UnsupportedOperationException()
+
+    private enum class FiltersState { NOT_FETCHED, FETCHING, FETCHED }
+
+    private inline fun <reified R> List<*>.firstInstanceOrNull(): R? =
+        filterIsInstance<R>().firstOrNull()
+
+    override fun setupPreferenceScreen(screen: PreferenceScreen) {
+        SwitchPreferenceCompat(screen.context).apply {
+            key = PREF_DYNAMIC_URL
+            title = PREF_DYNAMIC_URL_TITLE
+            summary = PREF_DYNAMIC_URL_SUMMARY
+            setDefaultValue(PREF_DYNAMIC_URL_DEFAULT)
+        }.let(screen::addPreference)
+    }
+
+    private var SharedPreferences.slugMap: MutableMap<String, String>
+        get() {
+            val jsonMap = getString(PREF_SLUG_MAP, "{}")!!
+            val slugMap = runCatching { json.decodeFromString<Map<String, String>>(jsonMap) }
+            return slugMap.getOrNull()?.toMutableMap() ?: mutableMapOf()
+        }
+        set(newSlugMap) {
+            edit()
+                .putString(PREF_SLUG_MAP, json.encodeToString(newSlugMap))
+                .apply()
+        }
+
+    private fun SharedPreferences.dynamicUrl(): Boolean = getBoolean(PREF_DYNAMIC_URL, PREF_DYNAMIC_URL_DEFAULT)
+
+    private fun String.toPermSlugIfNeeded(): String {
+        if (!preferences.dynamicUrl()) return this
+        val slug = this.substringAfter("/series/").substringBefore("/")
+        val absSlug = slug.substringBeforeLast("-")
+        preferences.slugMap = preferences.slugMap.apply { put(absSlug, slug) }
+        return this.replace(slug, absSlug)
+    }
+
+    companion object {
+        private val CLEAN_DATE_REGEX = """(\d+)(st|nd|rd|th)""".toRegex()
+        private val OLD_FORMAT_MANGA_REGEX = """^/manga/\d*(.*?)/?$""".toRegex()
```

Comment:
/manga/1223-cool-manga/
/manga/cool-manga/

```suggestion
        private val OLD_FORMAT_MANGA_REGEX = """^/manga/(\d+-)?([^/]+)/?$""".toRegex()
```

would have to adjust the match group elsewhere in the code too